lib/model/unreads.dart

```diff
@@ -0,0 +1,383 @@
+              ),
+            );
+
+          case UpdateMessageFlagsRemoveEvent():
+            mentions.removeAll(event.messages);
+        }
+
+      case MessageFlag.read:
+        switch (event) {
+          case UpdateMessageFlagsAddEvent():
+            if (event.all) {
+              streams.clear();
+              dms.clear();
+              mentions.clear();
+              oldUnreadsMissing = false;
+            } else {
+              mentions.removeAll(event.messages);
+              _slowRemoveAllInStreams(event.messages);
+              _slowRemoveAllInDms(event.messages);
+            }
+          case UpdateMessageFlagsRemoveEvent():
+            final newlyUnreadInStreams = <int, Map<String, QueueList<int>>>{};
+            final newlyUnreadInDms = <DmNarrow, QueueList<int>>{};
+            for (final messageId in event.messages) {
+              final detail = event.messageDetails![messageId];
+              if (detail == null) { // TODO(log) if on Zulip 6.0+
+                // Happens as a bug in some cases before fixed in Zulip 6.0:
+                //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.20in.20unsubscribed.20streams/near/1458467
+                // TODO(server-6) remove Zulip 6.0 comment
+                return;
+              }
+              switch (detail.type) {
+                case MessageType.stream:
+                  ((newlyUnreadInStreams[detail.streamId!] ??= {})[detail.topic!] ??= QueueList())
+                    .add(messageId);
+                case MessageType.private:
+                  final narrow = DmNarrow.ofUpdateMessageFlagsMessageDetail(selfUserId: selfUserId,
+                    detail);
+                  (newlyUnreadInDms[narrow] ??= QueueList())
+                    .add(messageId);
+              }
+            }
+            newlyUnreadInStreams.forEach((incomingStreamId, incomingTopics) {
+              incomingTopics.forEach((incomingTopic, incomingMessageIds) {
+                _addAllInStreamTopic(incomingMessageIds..sort(), incomingStreamId, incomingTopic);
+              });
+            });
+            newlyUnreadInDms.forEach((incomingDmNarrow, incomingMessageIds) {
+              _addAllInDm(incomingMessageIds..sort(), incomingDmNarrow);
+            });
+        }
+    }
+    notifyListeners();
+  }
+
+  // TODO replace with efficient alternative
+  bool _slowIsPresentInStreams(int messageId) {
+    return streams.values.any(
+      (topics) => topics.values.any(
+        (messageIds) => messageIds.contains(messageId),
+      ),
+    );
+  }
+
+  void _addLastInStreamTopic(int messageId, int streamId, String topic) {
+    ((streams[streamId] ??= {})[topic] ??= QueueList()).addLast(messageId);
+  }
+
+  /// [messageIds] must be sorted ascending and without duplicates.
+  void _addAllInStreamTopic(QueueList<int> messageIds, int streamId, String topic) {
```

Comment:
nit:
```suggestion
  // [messageIds] must be sorted ascending and without duplicates.
  void _addAllInStreamTopic(QueueList<int> messageIds, int streamId, String topic) {
```
because it doesn't have the form of a complete piece of documentation